rcl/src/rcl/node.c

```diff
@@ -536,6 +563,151 @@ rcl_get_disable_loaned_message(bool * disable_loaned_message)
   *disable_loaned_message = (strcmp(env_val, "1") == 0);
   return RCL_RET_OK;
 }
+
+static void rcl_node_type_cache_handle_service_call(const void *user_data,
+                                                    size_t number_of_events) {
+  (void)number_of_events;
+  rmw_request_id_t request_header;
+  type_description_interfaces__srv__GetTypeDescription_Request request;
+  type_description_interfaces__srv__GetTypeDescription_Response response;
+  rcl_node_type_cache_type_info_t type_info;
+
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+      user_data, "type description service callback: Passed user data is NULL",
+      return;);
+
+  const rcl_node_t *node = user_data;
+  const char *node_name = rcl_node_get_name(node);
+
+  if (!(type_description_interfaces__srv__GetTypeDescription_Request__init(
+            &request) &&
+        type_description_interfaces__srv__GetTypeDescription_Response__init(
+            &response))) {
+    RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME,
+                            "Failed to initialize service request / response.");
+    goto cleanup;
+  }
+
+  if (RCL_RET_OK != rcl_take_request(node->impl->get_type_description_service,
+                                     &request_header, &request)) {
+    RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Failed to take request");
+    goto cleanup;
+  }
+
+  // TODO(achim-k): Remove debug logging.
+  RCUTILS_LOG_INFO_NAMED(ROS_PACKAGE_NAME,
+                         "type_description service called for node '%s' with "
+                         "type version hash '%s'",
+                         node_name, request.type_hash.data);
+
+  rcl_ret_t ret = rcl_node_type_cache_get_type_info(
+      node, request.type_hash.data, &type_info);
+
+  if (RCUTILS_RET_OK == ret) {
+    // TODO(achim-k): Populate response
```

Comment:
This is an important bit. We need utility functions here in `rcl` that convert between `rosidl_runtime_c__type_description` <--> `type_description_interfaces__msg`. I think we could add those in this PR. Probably in a new header.  Something like this:

```
const type_description_msgs__msg__TypeDescription *
rcl_convert_type_description_runtime_to_msg(
  const rosidl_runtime_c__type_description__TypeDescription * runtime_description); 


rcl_convert_type_description_msg_to_runtime(
  const  type_description_msgs__msg__TypeDescription * description_msg, 
  const rosidl_runtime_c__type_description__TypeDescription ** out_runtime); 

// For users, convenience macro to fetch a description msg directly
#define RCL_GET_MSG_TYPE_DESCRIPTION(PkgName, MsgSubfolder, MsgName) \
    rcl_convert_type_description_runtime_to_msg(ROSIDL_GET_MSG_TYPE_SUPPORT(PkgName, MsgSubfolder, MsgName)->type_description)
```